addons/sale/controllers/portal.py

```diff
@@ -168,54 +169,64 @@ def portal_order_page(self, order_id, report_type=None, access_token=None, messa
         # Make sure that the partner's company matches the order's company.
         company_mismatch = not payment_portal.PaymentPortal._can_partner_pay_in_company(
-            order_sudo.partner_id, order_sudo.company_id
+            partner, company
         )
+        portal_page_values = {'company_mismatch': company_mismatch, 'expected_company': company}
+        providers_sudo = request.env['payment.provider'].sudo()._get_compatible_providers(
+            company.id,
+            partner.id,
+            amount,
+            currency_id=currency.id,
+            sale_order_id=order_sudo.id,
+            **kwargs
+        )  # In sudo mode to read the fields of providers and partner (if not logged in)
         fees_by_provider = {
             provider: provider._compute_fees(
-                order_sudo.amount_total,
-                order_sudo.currency_id,
-                order_sudo.partner_id.country_id,
+                amount, currency, partner.country_id
             ) for provider in providers_sudo.filtered('fees_active')
         }
-        portal_page_values = {
-            'company_mismatch': company_mismatch,
-            'expected_company': order_sudo.company_id,
-        }
         payment_form_values = {
             'providers': providers_sudo,
-            'tokens': tokens,
             'fees_by_provider': fees_by_provider,
-            'show_tokenize_input': PaymentPortal._compute_show_tokenize_input_mapping(
-                providers_sudo, logged_in=logged_in, sale_order_id=order_sudo.id
-            ),
-            'amount': order_sudo.amount_total,
-            'currency': order_sudo.currency_id,
-            'partner_id': order_sudo.partner_id.id,
-            'access_token': order_sudo.access_token,
+            'amount': amount,
+            'currency': currency,
+            'partner_id': partner.id,
+            'access_token': order_sudo._portal_ensure_token(),
             'transaction_route': order_sudo.get_portal_url(suffix='/transaction'),
             'landing_route': order_sudo.get_portal_url(),
         }
+        if tokens_included:
+            logged_in = not request.env.user._is_public()
+            show_tokenize_input = PaymentPortal._compute_show_tokenize_input_mapping(
+                providers_sudo, logged_in=logged_in, sale_order_id=order_sudo.id
+            )
+            children_domain = [('partner_id', 'child_of', partner.commercial_partner_id.id)]
+            partner_only_domain = [('partner_id', '=', partner.id)]
+            partner_domain = children_domain if 'child' in tokens_included else partner_only_domain
+            tokens = request.env['payment.token'].search(
+                expression.AND([partner_domain, [('provider_id', 'in', providers_sudo.ids)]])
+            ) if logged_in else request.env['payment.token']
+            payment_form_values.update({
+                'show_tokenize_input': show_tokenize_input,
+                'tokens': tokens,
```

Comment:
Is it ok that 'tokens' is not provided anymore when `tokens_included == None`?

I know that the simpler answer is "Yes, because it's only for the express checkout form", but I don't find it robust, as an API-like method. Say someone calls this method from a module that never shows tokens, they will probably pass `tokens_included=None`, which will not work because the payment form still expects `tokens` to be an iterable.